sled-agent/src/services.rs

```diff
@@ -274,42 +292,168 @@ impl Config {
-struct ZoneRequest {
-    zone: ServiceZoneRequest,
-    // TODO: Consider collapsing "root" into ServiceZoneRequest
+impl OmicronZonesConfigLocal {
+    pub fn to_omicron_zones_config(self) -> OmicronZonesConfig {
+        OmicronZonesConfig {
+            version: self.omicron_version,
+            zones: self.zones.into_iter().map(|z| z.zone).collect(),
+        }
+    }
+}
+
+/// Combines the Nexus-provided `OmicronZoneConfig` (which describes what Nexus
+/// wants for this zone) with any locally-determined configuration (like the
+/// path to the root filesystem)
+#[derive(
+    Clone, Debug, serde::Serialize, serde::Deserialize, schemars::JsonSchema,
+)]
+pub struct OmicronZoneConfigLocal {
+    pub zone: OmicronZoneConfig,
     #[schemars(with = "String")]
+    pub root: Utf8PathBuf,
+}
+
+/// Describes how we want a switch zone to be configured
+///
+/// This is analogous to `OmicronZoneConfig`, but for the switch zone (which is
+/// operated autonomously by the Sled Agent, not managed by Omicron).
+#[derive(Clone)]
+struct SwitchZoneConfig {
+    id: Uuid,
+    addresses: Vec<Ipv6Addr>,
+    services: Vec<SwitchService>,
+}
+
+/// Describes one of several services that may be deployed in a switch zone
+///
+/// Some of these are only present in certain configurations (e.g., with a real
+/// Tofino vs. SoftNPU) or are configured differently depending on the
+/// configuration.
+#[derive(Clone)]
+enum SwitchService {
+    ManagementGatewayService,
+    Wicketd { baseboard: Baseboard },
+    Dendrite { asic: DendriteAsic },
+    Tfport { pkt_source: String, asic: DendriteAsic },
+    Uplink,
+    MgDdm { mode: String },
+    Mgd,
+    SpSim,
+}
+
+impl crate::smf_helper::Service for SwitchService {
+    fn service_name(&self) -> String {
+        match self {
+            SwitchService::ManagementGatewayService => "mgs",
+            SwitchService::Wicketd { .. } => "wicketd",
+            SwitchService::Dendrite { .. } => "dendrite",
+            SwitchService::Tfport { .. } => "tfport",
+            SwitchService::Uplink { .. } => "uplink",
+            SwitchService::MgDdm { .. } => "mg-ddm",
+            SwitchService::Mgd => "mgd",
+            SwitchService::SpSim => "sp-sim",
+        }
+        .to_owned()
+    }
+    fn smf_name(&self) -> String {
+        format!("svc:/oxide/{}", self.service_name())
+    }
+    fn should_import(&self) -> bool {
+        true
+    }
+}
+
+/// Combines the generic `SwitchZoneConfig` with other locally-determined
+/// configuration
+///
+/// This is analogous to `OmicronZoneConfigLocal`, but for the switch zone.
+struct SwitchZoneConfigLocal {
+    zone: SwitchZoneConfig,
     root: Utf8PathBuf,
 }
 
+/// Describes either an Omicron-managed zone or the switch zone, used for
+/// functions that operate on either one or the other
+enum ZoneArgs<'a> {
+    Omicron(&'a OmicronZoneConfigLocal),
+    SledLocal(&'a SwitchZoneConfigLocal),
```

Comment:
Naming brainstorming: If we're discussing ownership, it might make sense to describe these as "Nexus-owned" vs "Sled-owned" services? I dunno if that helps, but it seems to be the distinction to me. (IMO the name "Omicron" for any zones doesn't actually provide extra context, since **all** these zones are kinda "Omicron zones").

Reply:
> Tiny nit / question - why `SledLocal` instead of something like `SwitchZone`? "Sled local" feels kind of vague (as does `sled_local_services()` below); feels like `Switch` or `SwitchZone` would be clearer.

Before this PR, the code had some notion of "sled-local zones" that I inferred was intended for the case where we have other kinds of non-Nexus-managed zones (hypothetically, in the future).  But it wasn't truly general: in practice it assumed in various places that a sled local zone _was_ the switch zone.  It's hard to avoid this since there was only one consumer so at some point where you _would_ write the code that matches on the zone type, you wind up just making the assumption that it is the one type of zone that it could be.  In this PR, that point is here (by calling this `SledLocal` but it contains a `SwitchZoneConfigLocal`).  We could call this `Switch` instead -- then that point would be in the caller.  (Or we'd eliminate the idea of sled-local zones altogether.  But the distinction does feel useful.)

> Naming brainstorming: If we're discussing ownership, it might make sense to describe these as "Nexus-owned" vs "Sled-owned" services? I dunno if that helps, but it seems to be the distinction to me. (IMO the name "Omicron" for any zones doesn't actually provide extra context, since all these zones are kinda "Omicron zones").

Yeah there's two axes here:

I went with "managed" (rather than "owned") because it better conveys (to me) that Nexus makes the high-level decisions about what zones run and what they're configuration is, but Sled Agent is responsible for the operation of the zone.  ("Nexus-owned" sounds to me that Nexus is fully responsible for this thing, which isn't true.)

On "Omicron" vs. "Nexus"...  "Nexus" is not quite right because these are also configured by RSS.  (In fact, they're exclusively configured by RSS right now and for the immediate future.)  I don't have a really strong feeling, and obviously we've used the term "Omicron" pretty generally to describe all kinds of stuff.  But it was initially defined to mean "the control plane", which I take to mean "the coordinated, distributed system that makes the decisions".  And I think that's what defines these zones -- that they're being managed by something external to Sled Agent, and that thing is the control plane [and we call that Omicron].  I hear the point about sled agent being part of the control plane, which is also true.  I don't have better ideas.

Reply:
> > Tiny nit / question - why `SledLocal` instead of something like `SwitchZone`? "Sled local" feels kind of vague (as does `sled_local_services()` below); feels like `Switch` or `SwitchZone` would be clearer.
> 
> Before this PR, the code had some notion of "sled-local zones" that I inferred was intended for the case where we have other kinds of non-Nexus-managed zones (hypothetically, in the future). But it wasn't truly general: in practice it assumed in various places that a sled local zone _was_ the switch zone. It's hard to avoid this since there was only one consumer so at some point where you _would_ write the code that matches on the zone type, you wind up just making the assumption that it is the one type of zone that it could be. In this PR, that point is here (by calling this `SledLocal` but it contains a `SwitchZoneConfigLocal`). We could call this `Switch` instead -- then that point would be in the caller. (Or we'd eliminate the idea of sled-local zones altogether. But the distinction does feel useful.)

Hmm. So if in the future we have a new `FooZone` that is managed by sled-agent like the switch zone today, would you envision that resulting in something like this?

```rust
enum ZoneArgs {
    Omicron(OmicronZoneConfigLocal),
    SledLocal(SledLocalZoneConfigLocal), // sorry for the terrible name; this is just a sketch
}

enum SledLocalZoneConfigLocal {
    Switch(SwitchZoneConfigLocal),
    Foo(FooZoneConfigLocal),
}
```

Hard to make this call in isolation, but I would guess that this extra level of enum isn't all that valuable, assuming there are methods that take any `ZoneArg`, methods that take switch zone args, and methods that take foo zone args, and there wouldn't be a need for methods that operate on "any sled-agent-managed zone but not omicron zones"? But maybe that's wrong. Sticking with it for now, I'd have proposed

```rust
enum ZoneArgs {
    Omicron(OmicronZoneConfigLocal),
    Switch(SwitchZoneConfigLocal),
    Foo(FooZoneConfigLocal),
}
```

I believe it's quite easy to change between at any point between now and the introduction of a real `Foo` zone, since it's all internal to sled-agent. I think I'd still vote for `ZoneArgs::Switch` for now, since it's the only sled-local zone we have for the foreseeable future.

Reply:
I think it would depend on the details (e.g., how similar the other kind of zone was).  I'm good with changing this to `Switch` now.

Reply:
Fixed

Reply:
> Naming brainstorming: If we're discussing ownership, it might make sense to describe these as "Nexus-owned" vs "Sled-owned" services? I dunno if that helps, but it seems to be the distinction to me. (IMO the name "Omicron" for any zones doesn't actually provide extra context, since **all** these zones are kinda "Omicron zones").

With apologies if this isn't adding much to what's already been said:

* My preference is for "managed" over "owned" for similar reasons to dap above
* I don't have a strong preference for Omicron-managed vs Nexus-managed. We don't really have a word for "the non-sled-agent control plane"; that _mainly_ means "Nexus" over the course of a rack's lifetime, but critically does mean "RSS" for major parts of sled-agent's codebase.

I'm tempted to propose `ControlPlaneManaged`, but that seems just as vague as (and basically synonymous with) `OmicronManaged`. I think in light of that I'm leaning toward `OmicronManaged`, unless we have (or can come up with) a name for "RSS-then-Nexus".